src/groupeddataframe/splitapplycombine.jl

```diff
@@ -766,49 +825,52 @@ function _combine(f::AbstractVector{<:Pair}, gd::GroupedDataFrame)
     if !all(x -> length(x) == length(outcols[1]), outcols)
         throw(ArgumentError("all functions must return values of the same length"))
     end
-    nams = [f[i] isa Pair{<:ColumnIndex} ?
-                Symbol(names(gd.parent)[index(gd.parent)[first(f[i])]],
-                       '_', funname(last(f[i]))) :
-                Symbol('x', i)
-            for i in 1:length(f)]
-    valscat = DataFrame(collect(AbstractVector, outcols), nams, makeunique=true)
+    return idx, DataFrame(collect(AbstractVector, outcols), nms)
+end
+
+function _combine(fun::Function, gd::GroupedDataFrame, ::Nothing)
+    idx, outcols, nms = _combine_with_first(wrap(fun(gd[1])), fun, gd, nothing)
+    valscat = DataFrame(collect(AbstractVector, outcols), nms)
     return idx, valscat
 end
 
-function _combine(f::Any, gd::GroupedDataFrame)
-    if f isa Pair{<:ColumnIndex}
-        incols = gd.parent[!, first(f)]
-        fun = last(f)
-    elseif f isa Pair
-        df = gd.parent[! , collect(first(f))]
-        incols = NamedTuple{Tuple(names(df))}(eachcol(df))
-        fun = last(f)
+function _combine(p::Pair, gd::GroupedDataFrame, ::Nothing)
+    f = normalize_selection(index(parent(gd)), p)
+    if first(f) isa Int
+        incols = parent(gd)[!, first(f)]
+        fun = first(last(f))
     else
-        incols = nothing
-        fun = f
+        incols = ntuple(i -> parent(gd)[!, first(f)[i]], length(first(f)))
+        fun = first(last(f))
     end
     agg = check_aggregate(fun)
-    if agg isa AbstractAggregate && f isa Pair && first(f) isa ColumnIndex
-        idx = Vector{Int}(undef, length(gd))
-        fillfirst!(nothing, idx, 1:length(gd.groups), gd)
+    if agg isa AbstractAggregate && first(f) isa Int
+        idx = gd.idx[gd.starts]
```

Comment:
I understand that `fillfirst!` is needed if you have more than one aggregate computation. In this branch we know that it is only one. But I can look into it in detail (I have not checked it, because it was your old code :smile:).

Reply:
OK, let me know if you understand it, otherwise I can try to understand why I wrote this.

Reply:
I got rid of needing to understand it in full by a different dispatch (though I am pretty sure it does what I have written)